src/main/flight/mixer.c

```diff
@@ -346,14 +346,92 @@ static void applyFlipOverAfterCrashModeToMotors(void)
     }
 }
 
+#ifdef USE_RPM_LIMITER
+static void applyRPMLimiter(void)
+{
+    if (mixerConfig()->rpm_limiter && motorConfig()->dev.useDshotTelemetry) {
+        float scaledRPMLimit = 0;
+        float pidOutput = 0;
+        float rcCommandThrottle = (rcCommand[THROTTLE] - 1000) / 1000.0f;
+        float averageRPM = 0;
+        float averageRPMSmoothed = 0;
+        bool motorsSaturated = false;
+        for (int i = 0; i < getMotorCount(); i++) {
+            averageRPM += getDshotTelemetry(i);
+            if (motor[i] >= motorConfig()->maxthrottle) {
+                motorsSaturated = true;
+            }
+        }
+        averageRPM = 100 * averageRPM / (getMotorCount() * motorConfig()->motorPoleCount / 2.0f);
+        averageRPMSmoothed = mixerRuntime.rpmLimiterPreviousSmoothedRPM + mixerRuntime.rpmLimiterDelayK * (averageRPM - mixerRuntime.rpmLimiterPreviousSmoothedRPM); //kinda braindead to convert to rps then back
+        if (mixerConfig()->rpm_limiter_rpm_linearization) {
+            scaledRPMLimit = ((mixerConfig()->rpm_limiter_rpm_limit - mixerConfig()->rpm_limiter_idle_rpm)) * 100.0f * (rcCommandThrottle) + mixerConfig()->rpm_limiter_idle_rpm * 100.0f;
+            float acceleration = scaledRPMLimit - mixerRuntime.rpmLimiterPreviousRPMLimit;
+            if (acceleration > 0) {
+                acceleration = MIN(acceleration, mixerRuntime.rpmLimiterAccelerationLimit);
+                scaledRPMLimit = mixerRuntime.rpmLimiterPreviousRPMLimit + acceleration;
+            }
+        } 
+        else {
+            throttle = throttle * mixerRuntime.rpmLimiterExpectedThrottleLimit;
+            scaledRPMLimit = ((mixerConfig()->rpm_limiter_rpm_limit)) * 100.0f;
```

Comment:
I do not understand this line of code, it seems wrong as it will always use the same number here for the scaledRPMLimit, which will create a very weird error when you compare it to the average motor output. Say you are at idle, this would always assume that there is a very large error...

Reply:
Line 376 is applying the dynamic throttle limit
Line 377:
You are correct that there would be a large error, but I limit the pid output to only be positive on line 406. This is because we only ever want the pid controller to reduce the throttle to bring rpms to the setpoint. The regulation is admittedly not the best, which is why I tested pre-activating the pid controller once the rpms got close to the limit, but this felt really weird. I think the dynamic throttle limit helps hide poor regulation, but this was also why I started experimenting with linearization which doesn't have this issue.

Reply:
> 

I guess I would need to see some logs of that method to see what its doing a bit better.